bindings/rust/aws-kms-tls-auth/src/receiver.rs

```diff
@@ -0,0 +1,358 @@
+    pub fn new(
+        client: Client,
+        trusted_key_arns: Vec<KeyArn>,
+        obfuscation_keys: Vec<ObfuscationKey>,
+    ) -> Self {
+        let key_cache = moka::sync::Cache::builder()
+            .max_capacity(MAXIMUM_KEY_CACHE_SIZE as u64)
+            .time_to_live(KEY_ROTATION_PERIOD * 2)
+            .build();
+        Self {
+            client,
+            trusted_key_arns: Arc::new(trusted_key_arns),
+            obfuscation_keys,
+            key_cache,
+        }
+    }
+
+    /// This is the main async future that s2n-tls polls.
+    ///
+    /// It will
+    /// 1. decrypt the ciphertext datakey
+    /// 2. check that the decrypted material is associated with a trusted key id
+    /// 3. cache the decrypted material in the key cache
+    /// 4. return an s2n-tls psk
+    ///
+    /// All of the arguments are owned to satisfy the `'static` bound that s2n-tls
+    /// requires on connection futures.
+    async fn kms_decrypt_and_update(
+        psk_identity: Vec<u8>,
+        ciphertext_datakey: Vec<u8>,
+        client: Client,
+        trusted_key_arns: Arc<Vec<KeyArn>>,
+        key_cache: Cache<Vec<u8>, Vec<u8>>,
+    ) -> anyhow::Result<s2n_tls::psk::Psk> {
+        let ciphertext_datakey_clone = ciphertext_datakey.clone();
+        let decrypted = tokio::spawn(async move {
+            client
+                .decrypt()
+                .ciphertext_blob(Blob::new(ciphertext_datakey_clone))
+                .send()
+                .await
+        })
+        .await??;
+
+        // although the field is called `key_id`, it is actually the key arn. This
+        // is confirmed in the documentation:
+        // https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html#API_Decrypt_ResponseSyntax
+        let associated_key_arn = decrypted.key_id.as_ref().unwrap();
+        if !trusted_key_arns.contains(associated_key_arn) {
+            anyhow::bail!("untrusted KMS Key: {associated_key_arn} is not trusted");
+        }
+
+        let plaintext_datakey = decrypted.plaintext.unwrap().into_inner();
+
+        key_cache.insert(ciphertext_datakey, plaintext_datakey.clone());
+
+        let psk = psk_from_material(&psk_identity, &plaintext_datakey).unwrap();
+
+        Ok(psk)
+    }
+}
+
+impl ClientHelloCallback for PskReceiver {
+    fn on_client_hello(
+        &self,
+        connection: &mut s2n_tls::connection::Connection,
+    ) -> Result<Option<Pin<Box<dyn ConnectionFuture>>>, s2n_tls::error::Error> {
+        // parse the identity list from the client hello
+        let client_hello = connection.client_hello()?;
+        let identities = match retrieve_psk_identities(client_hello) {
+            Ok(identities) => identities,
+            Err(e) => {
+                return Err(s2n_tls::error::Error::application(e.into()));
+            }
+        };
+
+        // extract the identity bytes
+        let psk_identity = match identities.list().first() {
```

Comment:
Is this an intentional simplification? We assume we're talking to a PskProvider, so there will only be one PSK?